rtmr/rtmr.go

```diff
@@ -0,0 +1,210 @@
+	tsmPathTcgMap = "tcg_map"
+)
+
+// Extend is a struct that represents a rtmr entry in the configfs.
+type Extend struct {
+	RtmrIndex int
+	entry     *configfsi.TsmPath
+	client    configfsi.Client
+}
+
+// Response is a struct that represents the response of reading a rtmr entry in the configfs.
+type Response struct {
+	RtmrIndex int
+	digest    []byte
+	tcgMap    []byte
+}
+
+func (r *Extend) attribute(subtree string) string {
+	a := *r.entry
+	a.Attribute = subtree
+	return a.String()
+}
+
+// extendDigest extends the measurement to the rtmr with the given hash.
+func (r *Extend) extendDigest(hash []byte) error {
+	if err := r.client.WriteFile(r.attribute(tsmRtmrDigest), hash); err != nil {
+		return fmt.Errorf("could not write digest to rmtr%d: %v", r.RtmrIndex, err)
+	}
+	return nil
+}
+
+// getDigest returns the digest of the rtmr.
+func (r *Extend) getDigest() ([]byte, error) {
+	return r.client.ReadFile(r.attribute(tsmRtmrDigest))
+}
+
+// getTcgMap returns the tcg map of the rtmr.
+func (r *Extend) getTcgMap() ([]byte, error) {
+	return r.client.ReadFile(r.attribute(tsmPathTcgMap))
+}
+
+// validateIndex checks if the rtmr index is valid.
+func (r *Extend) validateIndex() error {
+	indexBytes, err := r.client.ReadFile(r.attribute(tsmPathIndex))
+	if err != nil {
+		return fmt.Errorf("could not read index %s: %v", r.attribute(tsmPathIndex), err)
+	}
+	index, err := configfsi.Kstrtouint(indexBytes, 10, 64)
+	if err != nil {
+		return fmt.Errorf("could not convert index %s to integer: %v", indexBytes, err)
+	}
+	if (int)(index) != r.RtmrIndex {
+		return fmt.Errorf("rtmr index %d does not match the expected index %d", index, r.RtmrIndex)
+	}
+	return nil
+}
+
+// setRtmrIndex sets a configfs rtmr entry to the given index.
+// It reports an error if the index cannot be written.
+func (r *Extend) setRtmrIndex() error {
+	indexBytes := []byte(strconv.Itoa(r.RtmrIndex)) // Convert index to []byte
+	indexPath := r.attribute(tsmPathIndex)
+	if err := r.client.WriteFile(indexPath, indexBytes); err != nil {
+		return fmt.Errorf("could not write index %s: %v", indexPath, err)
+	}
+	return nil
+}
+
+// searchRtmrInterface searches for an rtmr entry in the configfs.
+func searchRtmrInterface(client configfsi.Client, index int) (*Extend, error) {
+	root := tsmRtmrPrefix
+	out := &Extend{}
+	if err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() {
+			p, err := configfsi.ParseTsmPath(path)
+			if err != nil {
+				return err
+			}
+			r := &Extend{
+				RtmrIndex: index,
+				entry:     &configfsi.TsmPath{Subsystem: rtmrSubsystem, Entry: p.Entry},
+				client:    client,
+			}
+			if r.validateIndex() == nil {
+				out = r
+				return nil
+			}
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := out.validateIndex(); err != nil {
+		return out, nil
+	} else {
+		return nil, err
+	}
+
```

Comment:
This can be written as `return out, out.validateIndex()` since you shouldn't read the result if the error is non-nil. I would also like to see a test for what this looks like when the walk does not find an entry.